src/Symfony/Bridge/Twig/Extension/TranslationExtension.php

```diff
@@ -104,17 +104,28 @@ public function getTranslationNodeVisitor(): TranslationNodeVisitor
     }
 
     /**
-     * @param ?string|Translatable $message The message id (may also be an object that can be cast to string)
+     * @param string|\Stringable|TranslatableInterface|null $message
+     * @param array|string                                  $arguments Can be the locale as a string when $message is a TranslatableInterface
      */
-    public function trans($message, array $arguments = [], string $domain = null, string $locale = null, int $count = null): string
+    public function trans($message, $arguments = [], string $domain = null, string $locale = null, int $count = null): string
     {
-        if ($message instanceof Translatable) {
-            $arguments += $message->getParameters();
-            $domain = $message->getDomain();
-            $message = $message->getMessage();
+        if ($message instanceof TranslatableInterface) {
+            if (1 < \func_num_args() && !\is_string($arguments)) {
+                throw new \TypeError(sprintf('Argument 2 passed to "%s()" must be a locale passed as a string when the message is a "%s", "%s" given.', __METHOD__, TranslatableInterface::class, get_debug_type($arguments)));
+            }
+
+            if (null !== ($domain ?? $locale ?? $count)) {
+                throw new \ArgumentCountError(sprintf('Method "%s()" accepts max two arguments when the message is a "%s".', __METHOD__, TranslatableInterface::class));
```

Comment:
This logic won't play well with Twig named arguments, which allow to pass `locale='fr'` which will then call `->trans($message, [], null, 'fr')` in the compiled code.

Reply:
Thanks, updated, can you have a second look please?

Reply:
I just thought about another incompatibility. I'm quite sure that this strict argument validation will make Translatable incompatible with templates using `{% trans_default_domain %}` (or maybe compatible but only if using `transobject|trans(domain=null)` which would be really weird, but that's untested), due to `{% trans_default_domain %}` injecting the `$domain` argument automatically anytime the call does not have it yet.

Maybe we should not try to be that strict about argument validation.

Reply:
good catch, I removed this check now